Flatpak fixes

hadess opened the pull request:
This makes it possible to easily rip a DVD on a Linux system, and updates the Flatpak manifest to close a few sandbox holes.
If the libdvdcss commit isn't acceptable, I'm happy to remove it.

hadess committed:
LinGui: detect DVD drives without udev

This is a good fallback when udev isn't available, such as within a
Flatpak sandbox.

hadess committed:
flatpak: Allow accessing optical drives

hadess commented:
Is there an easy way to actually build the modified flatpak manifest locally? It seems to rely on `handbrake.tar.bz2` and `download.tgz` being in the source directory, and I don't know how to generate those from modified sources.

sr55 commented:
We no longer support libdvdcss so wouldn't accept any patches adding support for it.

Anything else, I'll leave for john to review.

bradleysepos commented:
John put together this article awhile back, not sure if it helps: https://handbrake.fr/docs/en/latest/developer/flatpak-repo.html

hadess commented:
> We no longer support libdvdcss so wouldn't accept any patches adding support for it.

I don't understand what that means. How does Handbrake decrypt DVDs on Linux? Does it use VLC like on macOS? How does that work when vlc isn't available (like in a Flatpak)?

hadess commented:
> John put together this article awhile back, not sure if it helps: https://handbrake.fr/docs/en/latest/developer/flatpak-repo.html

It's helpful, and not really at the same time. I don't understand why I need to install nasm, some numa libraries and a bunch of other packages on my host system when the whole application will be built in a sandbox with a different runtime.

I modified the manifest to point directly at my local git repo, and it tried to download a whole bunch of libraries from the network (which it doesn't have access to), including libdvdread, which at least answers my [question above](https://github.com/HandBrake/HandBrake/pull/3164#issuecomment-701604868).

So I can't really test those changes easily, but let me know if you run into something.

hadess commented:
FYI, so that HandBrake doesn't need to ship it itself: https://github.com/flathub/flathub/pull/1863

woodstockathbf commented:
Handbrake is not intended as ripping software. We're coming up on the twelfth anniversary of officially NOT supporting third-party libraries necessary to do that. As for "how does it rip on Linux/MacOS/Windows?", it _doesn't._

Forcing it to work can be done, and a lot of web pages out there tell you how to make it work, but it still doesn't fix the fact that the third-party library isn't a complete solution for removing DVD copy protection, and does nothing if you decide you want to rip Blu-ray and UHD disks. And even updating your operating system can cause it to stop working, which has happened many times.

Software that advertises that it does ripping is recommended, because their authors spend a LOT of time keeping up with the copy-protection wars.

hadess commented:
> Handbrake is not intended as ripping software. We're coming up on the twelfth anniversary of officially NOT supporting third-party libraries necessary to do that. As for "how does it rip on Linux/MacOS/Windows?", it _doesn't._

It's really weird for HandBrake to advertise and list DVD drives as sources but not support getting the data off of them (right now, it silently fails to read anything off the disc, as if it was broken).

> Software that advertises that it does ripping is recommended, because their authors spend a LOT of time keeping up with the copy-protection wars.

I think we can say that DVD ripping is well understood at this point. libdvdcss is close to 20 years old, and I doubt it's any major changes to the way it works. If ripping from DVDs isn't supported, it should be shown through the UI not advertising using DVDs as a source.

woodstockathbf commented:
If you read the documentation for handbrake, the [About](https://handbrake.fr/docs/en/1.3.0/introduction/about.html) page explains that it _does not remove copy protection_. There are DVDs and BDs out there without copy protection.

As for "DVD ripping is well understood at this point", removing the encryption is but one step, and that's all that libdvdcss does. DVDs have continued to evolve since that came out. Ask anyone who has ripped a DVD published by Lionsgate, where you have to choose which of 8 or more titles is correct (over a hundred on BDs), and some of the fake titles include information that will make your drive seek to impossible locations.

I agree with the handbrake developers' decision to not spend **their** time running down those rabbit holes. The developer of MakeMKV, for example, has spent over a year working on a generalized solution to a copy protection scheme (AACS 2.1) that, so far, affects less than 10 titles world-wide, and developing firmware patches to remove hardware-based copy protection that affects DVD, BD, and UHD. That's been going on longer than the gap between several point releases on handbrake, which I think are more important.

hadess commented:
> If you read the documentation for handbrake, the [About](https://handbrake.fr/docs/en/1.3.0/introduction/about.html) page explains that it _does not remove copy protection_. There are DVDs and BDs out there without copy protection.

There are so few DVDs without copy protection that you can't really be claiming to support DVDs if you don't have a way to remove that copy protection. I completely understand not wanting to remove copy protection for *reasons* (I've maintain a movie player that can play DVDs for close to 18 years, and I know the problems that can lead to, which is why I mentioned not bundling libdvdcss straight away), but the UI makes it look like it supports DVDs (*all* DVDs), doesn't throw any errors when DVDs *are* encrypted, and will happily handle encrypted DVDs when the magic library is installed.

If nothing else, I'd expect an error message saying that encrypted DVDs aren't supported if HandBrake can't access them without errors. For which all the fixes above would still be necessary.

You also keep dragging the discussion to Blu-rays, but this isn't what this PR is about.

> As for "DVD ripping is well understood at this point", removing the encryption is but one step, and that's all that libdvdcss does. DVDs have continued to evolve since that came out. Ask anyone who has ripped a DVD published by Lionsgate, where you have to choose which of 8 or more titles is correct (over a hundred on BDs), and some of the fake titles include information that will make your drive seek to impossible locations.

But HandBrake supports unencrypted discs, so you'd run into the same problems after having removed the copy protection, as that's not what's going to handle the layout. And I still didn't mention Blu-rays :)

woodstockathbf commented:
You don't mention BDs, but others have, based on "but HB does DVDs!" And, like DVDs, you CAN force handbrake to decode BDs using external libraries, as well, since it has a library for reading decrypted BD structures. That library can use lists of encryption keys to decode a BD. And people have replaced that library with links to MakeMKV or AnyDVD to use their decryption engines... which is still unsupported by the handbrake crew.

The GUI does not present an error message if you try to read an encrypted DVD, probably because the parts of the DVD structure the GUI needs to read are not encrypted, even on a commercial disk. I can see a good argument for putting up a warning when the DVD is loaded about "encryption detected, no support for decryption found, errors will follow".

The encode logs do show errors, though, once things are turned over to hb.dll and the actual video needs to be read. But, if you hang around here much, you know users do not read logs. :)

hadess commented:
> And, like DVDs, you CAN force handbrake to decode BDs using external libraries

It's a libdvdread feature.

> The GUI does not present an error message if you try to read an encrypted DVD, probably because the parts of the DVD structure the GUI needs to read are not encrypted, even on a commercial disk.

When I say that it fails silently, it fails silently. It can't parse the titles, even if the IFOs aren't encrypted, it just acts as if nothing was on the disc.

> hb.dll

That doesn't look like a Linux library... This is a UI problem with the GTK/Linux front-end. It has nothing to do with Windows.

bradleysepos commented:
> There are so few DVDs without copy protection that you can't really be claiming to support DVDs if you don't have a way to remove that copy protection.

I politely and categorically disagree with this statement. DVD is not simply commercial films and TV. There are tons of family memories, personal and commercial projects, film archives, etc. on DVD without copy protection. CSS and other protections/breakage do not define DVD. Same for BD.

As was previously mentioned, the project has made it very clear for more than a decade there is zero copy protection support, so it sounds like you have an expectation beyond what is explicitly stated.

In sum, DVD != CSS and BD != AACS. HandBrake does not and will not remove or work around any copy protection, period.

We appreciate the contributions, but let's not rehash a years old debate. We're a small team and could use fewer distractions.

hadess commented:
> In sum, DVD != CSS and BD != AACS. HandBrake does not and will not remove or work around any copy protection, period.
> 
> We appreciate the contributions, but let's not rehash a years old debate. We're a small team and could use fewer distractions.

There isn't anything related to copy protection in this PR, and I had already removed the libdvdcss module from the flatpak manifest before anyone started this discussion about copy protection.

jstebbins commented:
@hadess it looks like you are using icon properties for the source media to indentify the media type.  Cute, but kind of hacky. I haven't really looked at that part of the code in years.  At that time, gudev was the most expeditious and reliable way to detect if the source media was a DVD.  Is gudev not part of the flatpak gnome platform?  Or does it not work correctly inside flatpak sandbox?

Also, putting this new code after the ```#else``` wouldn't be enough.  Would have to be ```#elif defined(__linux__)``` and another ```#else``` added to the end.  That code is used for mingw as well.

Regarding tightening up the sandbox, I'm not sure what the effect of the user config dir changes are that you've made.  TBH, I originally copied those attributes from another project.  I assume that using the users home config allows things like matching any theme overrides they might be using, etc.

Changing ```device=all``` is the opposite of tightening and we should not allow this.  It grants access to all of ```/dev```. If the user wishes to process physical media, they can use ```flatpak override``` to grant the necessary permissions.

hadess commented:
> @hadess it looks like you are using icon properties for the source media to indentify the media type. Cute, but kind of hacky. I haven't really looked at that part of the code in years. At that time, gudev was the most expeditious and reliable way to detect if the source media was a DVD. Is gudev not part of the flatpak gnome platform? Or does it not work correctly inside flatpak sandbox?

libgudev can be built as a module, and it won't be able to talk to systemd-udevd, which lives outside the sandbox, on purpose. The other way to do this would be to look for the GDrive containing GVolumes containing a GMount, which would have the `x-content-type` for DVDs. You can see this in `gio mount -li` run outside the sandbox.

totem-pl-parser's disc (and image) identification code does that:
https://gitlab.gnome.org/GNOME/totem-pl-parser/-/blob/master/plparse/totem-disc.c

> Also, putting this new code after the `#else` wouldn't be enough. Would have to be `#elif defined(__linux__)` and another `#else` added to the end. That code is used for mingw as well.

Ha, I can certainly do that. Do you prefer that I do the above though? It will be a fair bit more work, but it's doable.

> Regarding tightening up the sandbox, I'm not sure what the effect of the user config dir changes are that you've made. TBH, I originally copied those attributes from another project. I assume that using the users home config allows things like matching any theme overrides they might be using, etc.

Themes, etc. are already proxied through xdg-desktop-portal and GTK (GtkSettings has a backend for outside, and inside the sandbox).

> Changing `device=all` is the opposite of tightening and we should not allow this. It grants access to all of `/dev`. If the user wishes to process physical media, they can use `flatpak override` to grant the necessary permissions.

That's currently the only way to get access to block devices, and that's also what totem (the GNOME movie player), or VLC use to get access to optical discs. Expecting end-users to use `flatpak override` isn't very user-friendly, and it's (part of) what ended up pushing me to use RPMFusion's version of HandBrake instead. It just looks broken as no CD devices are listed.

Note that HandBrake still only gets access to the same devices the user gets access to outside the sandbox. It's not like they can suddenly dd to fixed disks, or something.

This is the issue that tracks this:
https://github.com/flatpak/xdg-desktop-portal/issues/228

jstebbins commented:
> Ha, I can certainly do that. Do you prefer that I do the above though? It will be a fair bit more work, but it's doable

Yes please, it sounds like the Right Way:tm:.  In fact, if it works we should rip out gudev and just use the new and better method.  This is the only thing gudev is used for in ghb and would be nice to remove a library dependency.

> Themes, etc. are already proxied through xdg-desktop-portal and GTK (GtkSettings has a backend for outside, and inside the sandbox).

OK.  As I said, I copied that from another project when flatpak was still in early development.  It may have been needed back then, or may not have been needed at all.  If some side effect appears, we'll chase it down when it happens.

I forgot to ask earlier, what is the effect of ```fallback-x11``` vs ```x11```?

> That's currently the only way to get access to block devices

Yes I know.  I've seen discussions about a device portal since I first added flatpak packaging to HandBrake 4 years ago.  Unfortunately still no progress.  But since we have strongly recommended against this workflow for a dozen years or more, I think it's reasonable to ask the user to override permissions to access this feature.

Even DVDs created for personal use (i.e. home movies) almost always need to be preprocessed by ripping software that specializes in the kinds of problems they can have. In addition to CSS encryption,  more often than not, personal DVD creation software generates output that is not compliant to DVD specifications.  HandBrake shits the bed on most of these home generated DVDs and we really have no motivation to fix such issues since there is software available that specializes in this and will always do a better job than we will.

As a aside, I also lazily maintain a HandBrake plugin that allows reading BDs using MakeMKV's library that does decryption.  I haven't published it anywhere, it just sits on one of my computers. But it's the reason I created the flatpak plugin interface for HandBrake.  I did it more as a science experiment, never really intended to use it.  I always run everything through MakeMKV first  (which has it's own flatpak package https://flathub.org/apps/details/com.makemkv.MakeMKV).

hadess committed:
Remove udev dependency

In addition to udev not working correctly inside a sandbox, the icon
names given out by the gio library (through gvfs' udisks2 volume monitor)
are just as accurate as the udev optical disc detection, as gvfs uses
the icon name given by udisks itself, which derives them from the device
property udev adds:
https://github.com/storaged-project/udisks/blob/master/udisks/udisksobjectinfo.c#L177-L196
https://github.com/systemd/systemd/blob/master/src/udev/cdrom_id/cdrom_id.c#L914-L997

hadess commented:
> > Ha, I can certainly do that. Do you prefer that I do the above though? It will be a fair bit more work, but it's doable
> 
> Yes please, it sounds like the Right Waytm. In fact, if it works we should rip out gudev and just use the new and better method. This is the only thing gudev is used for in ghb and would be nice to remove a library dependency.

Unfortunately, I don't think that there's a much better way than this to detect the drives for now.
https://gitlab.gnome.org/GNOME/glib/-/issues/2230

I've added a new commit to remove the udev dependency though, as I think the current code is, even if hacky, as accurate as it's going to get for now.

> > Themes, etc. are already proxied through xdg-desktop-portal and GTK (GtkSettings has a backend for outside, and inside the sandbox).
> 
> OK. As I said, I copied that from another project when flatpak was still in early development. It may have been needed back then, or may not have been needed at all. If some side effect appears, we'll chase it down when it happens.

This was added in GTK 3.24.2 in 2018.

> I forgot to ask earlier, what is the effect of `fallback-x11` vs `x11`?

It stops Wayland applications on Wayland from being able to access the X11 socket, eg. the XWayland server. X11 applications all share the same connection to the X11 server, but Wayland apps don't need to talk to X11 apps.

> > That's currently the only way to get access to block devices
> 
> Yes I know. I've seen discussions about a device portal since I first added flatpak packaging to HandBrake 4 years ago. Unfortunately still no progress. But since we have strongly recommended against this workflow for a dozen years or more, I think it's reasonable to ask the user to override permissions to access this feature.

I still don't think it's a reasonable thing to ask of users if there's going to be UI that advertises those devices. I'd rather security conscious users override it the other way, or use Flatseal, rather than the disc import failing for users with weird error messages.

sr55 commented:
@jstebbins  @robxnano   Anything here still useful or do we close this out now?

robxnano commented:
The flatpak manifest changes have already been implemented elsewhere. The change to DVD handling still seems to work and would be preferable to the current method, but I think jstebbins suggested that the method @hadess described [here](https://github.com/HandBrake/HandBrake/pull/3164#issuecomment-702206964) would be preferable. If that's still case it would be easier to just close this one and create a new one with the alternative method, if hadess is still interested. Otherwise I could try implementing it.

jstebbins commented:
As @robxnano points out, the manifest changes that were accepted have been implemented.

The majority of the changes in this PR are the removal of gudev which relates to the change in DVD physical media detection. The GVolume method would be preferable, but I haven't investigated whether that method actually works in a flatpak. I just assumed it did because @hadess said it does. I would prefer that method over the once-removed icon properties detection method. But since I'm so uninvolved lately, I'm not going to insist.

hadess commented:
> The flatpak manifest changes have already been implemented elsewhere.

Really disappointed that instead of asking me to split off some of the changes in another PR, or cherry-picking the patches from this branch, the manifest changes were reimplemented a year later without any credit.

I've rebased the patches, do with them what you will.

sr55 commented:
@hadess  Apologies about that. This one fell though the cracks and a change came in independently of this. 

@robxnano  This has been rebased.  Being more of a gtk gui expert than I, mind doing a quick review?

If it's all good, I can merge.

robxnano commented on gtk/src/callbacks.c:
```diff
@@ -5122,50 +5118,31 @@ dvd_device_list (void)
-#if defined(__linux__) && defined(_HAVE_GUDEV)
-    udev_ctx = g_udev_client_new(NULL);
+    return ret;
 #endif
```
A stray #endif left over from the rebase.

robxnano reviewed (changes requested):
Just one problem, apart from that it all works fine.

galad87 commented:
Can this be merged?

sr55 commented:
Yup, think we are good now

sr55 commented:
Long overdue so apologies for that, but now merged.
